frame/contracts/src/lib.rs

```diff
@@ -949,10 +905,102 @@ pub mod pallet {
-		StorageValue<_, BoundedVec<DeletedContract, T::DeletionQueueDepth>, ValueQuery>;
+	pub(crate) type DeletionQueueNonce<T: Config> =
+		StorageValue<_, DeletionQueueNonces, ValueQuery>;
+}
+
+/// DeletionQueueNonces is a pair of monotonic counters used to index contracts marked for deletion,
+/// and the latest deleted contract.
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, Clone, Copy, Default)]
+struct DeletionQueueNonces {
+	/// Monotonic counter used as a key for inserting new deleted contract in the DeletionQueueMap.
+	/// The nonce is incremented after each insertion.
+	insert_nonce: u32,
+	/// The index used to read the next element to be deleted in the DeletionQueueMap.
+	/// The nonce is incremented after each deletion.
+	delete_nonce: u32,
+}
+
+impl DeletionQueueNonces {
+	/// Increment the delete nonce and return the new value.
+	fn delete_inc(&mut self) -> &mut Self {
+		self.delete_nonce = self.delete_nonce.wrapping_add(1);
+		self
+	}
+
+	/// Increment the insert nonce and return the new value.
+	fn insert_inc(&mut self) -> &mut Self {
+		self.insert_nonce = self.insert_nonce.wrapping_add(1);
+		self
+	}
+}
+
+/// DeletionQueue manage the removal of contracts storage that are marked for deletion.
+///
+/// When a contract is deleted by calling `seal_terminate` it becomes inaccessible
+/// immediately, but the deletion of the storage items it has accumulated is performed
+/// later by pulling the contract from the deletion queue in the `on_idle` hook.
+struct DeletionQueue<T: Config> {
+	nonces: DeletionQueueNonces,
+	_phantom: PhantomData<T>,
+}
+
+/// View on a contract that is marked for deletion
+/// The struct takes a mutable reference on the deletion queue so that the contract can be removed,
+/// and none can be added or read in the meantime.
+struct DeletionQueueEntry<'a, T: Config> {
+	contract: DeletedContract,
+	queue: &'a mut DeletionQueue<T>,
+}
+
+impl<'a, T: Config> DeletionQueueEntry<'a, T> {
+	/// Get a reference to the contract that is marked for deletion.
+	fn contract(&self) -> &DeletedContract {
+		&self.contract
+	}
+
+	/// Remove the contract from the deletion queue.
+	fn remove(self) {
+		<DeletionQueueMap<T>>::remove(self.queue.nonces.delete_nonce);
+		<DeletionQueueNonce<T>>::set(*self.queue.nonces.delete_inc());
+	}
+}
+
+impl<T: Config> DeletionQueue<T> {
+	/// Load the Deletion Queue nonces, so we can perform read or write operations on the
+	/// DeletionQueueMap storage
+	fn load() -> Self {
+		let nonces = <DeletionQueueNonce<T>>::get();
+		Self { nonces, _phantom: PhantomData }
+	}
+
+	/// The number of contracts marked for deletion.
+	fn len(&self) -> u32 {
+		self.nonces.insert_nonce.wrapping_sub(self.nonces.delete_nonce)
+	}
+
+	/// Insert a contract in the deletion queue.
+	fn insert(&mut self, contract: DeletedContract) {
+		<DeletionQueueMap<T>>::insert(self.nonces.insert_nonce, contract);
+		<DeletionQueueNonce<T>>::set(*self.nonces.insert_inc());
+	}
```

Comment:
We still need to check that the queue doesn't overflow here. Or you just go for `u64` and drop the ring buffer logic. We can then just assume that it never overflows. We do the same for the instantiation nonce.

Reply:
to overflow we would need u32::MAX pending value to be deleted, at which point we would start overwriting values 
and leak some storage. 

But otherwise this should be fine,  since we use wrapping_inc to increment the counters we will just start over from 0 once we get passed u32:::MAX

Reply:
added a test in this [commit](  https://github.com/paritytech/substrate/pull/13702/commits/737a48a2b9f35381ea55dcaef198535bf0da47da)